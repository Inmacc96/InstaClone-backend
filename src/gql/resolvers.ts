const resolvers = {
  Query: {
    // User
    getUser: () => {
      console.log("Getting user");
      return null;
    },
  },
  Mutation: {
    // User
    createUser: () => {
      console.log("Creating user...");
      return null;
    },
  },
};

export default resolvers;
