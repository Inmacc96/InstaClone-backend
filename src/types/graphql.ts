import { GraphQLResolveInfo } from 'graphql';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type RequireFields<T, K extends keyof T> = Omit<T, K> & { [P in K]-?: NonNullable<T[P]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type AuthInput = {
  email: Scalars['String'];
  password: Scalars['String'];
};

export type Mutation = {
  __typename?: 'Mutation';
  authUser: Token;
  newUser?: Maybe<User>;
};


export type MutationAuthUserArgs = {
  input: AuthInput;
};


export type MutationNewUserArgs = {
  input: UserInput;
};

export type Query = {
  __typename?: 'Query';
  getUser?: Maybe<User>;
};

export type Token = {
  __typename?: 'Token';
  token?: Maybe<Scalars['String']>;
};

export type User = {
  __typename?: 'User';
  avatar?: Maybe<Scalars['String']>;
  createdAt?: Maybe<Scalars['String']>;
  description?: Maybe<Scalars['String']>;
  email: Scalars['String'];
  id: Scalars['ID'];
  name: Scalars['String'];
  password: Scalars['String'];
  siteWeb?: Maybe<Scalars['String']>;
  username: Scalars['String'];
};

export type UserInput = {
  email: Scalars['String'];
  name: Scalars['String'];
  password: Scalars['String'];
  username: Scalars['String'];
};



export type ResolverTypeWrapper<T> = Promise<T> | T;


export type ResolverWithResolve<TResult, TParent, TContext, TArgs> = {
  resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | ResolverWithResolve<TResult, TParent, TContext, TArgs>;

export type ResolverFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => Promise<TResult> | TResult;

export type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => AsyncIterable<TResult> | Promise<AsyncIterable<TResult>>;

export type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<{ [key in TKey]: TResult }, TParent, TContext, TArgs>;
  resolve?: SubscriptionResolveFn<TResult, { [key in TKey]: TResult }, TContext, TArgs>;
}

export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
  resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}

export type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> =
  | SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs>
  | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;

export type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> =
  | ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>)
  | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;

export type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (
  parent: TParent,
  context: TContext,
  info: GraphQLResolveInfo
) => Maybe<TTypes> | Promise<Maybe<TTypes>>;

export type IsTypeOfResolverFn<T = {}, TContext = {}> = (obj: T, context: TContext, info: GraphQLResolveInfo) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (
  next: NextResolverFn<TResult>,
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = {
  AuthInput: AuthInput;
  Boolean: ResolverTypeWrapper<Scalars['Boolean']>;
  ID: ResolverTypeWrapper<Scalars['ID']>;
  Mutation: ResolverTypeWrapper<{}>;
  Query: ResolverTypeWrapper<{}>;
  String: ResolverTypeWrapper<Scalars['String']>;
  Token: ResolverTypeWrapper<Token>;
  User: ResolverTypeWrapper<User>;
  UserInput: UserInput;
};

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = {
  AuthInput: AuthInput;
  Boolean: Scalars['Boolean'];
  ID: Scalars['ID'];
  Mutation: {};
  Query: {};
  String: Scalars['String'];
  Token: Token;
  User: User;
  UserInput: UserInput;
};

export type MutationResolvers<ContextType = any, ParentType extends ResolversParentTypes['Mutation'] = ResolversParentTypes['Mutation']> = {
  authUser?: Resolver<ResolversTypes['Token'], ParentType, ContextType, RequireFields<MutationAuthUserArgs, 'input'>>;
  newUser?: Resolver<Maybe<ResolversTypes['User']>, ParentType, ContextType, RequireFields<MutationNewUserArgs, 'input'>>;
};

export type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
  getUser?: Resolver<Maybe<ResolversTypes['User']>, ParentType, ContextType>;
};

export type TokenResolvers<ContextType = any, ParentType extends ResolversParentTypes['Token'] = ResolversParentTypes['Token']> = {
  token?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type UserResolvers<ContextType = any, ParentType extends ResolversParentTypes['User'] = ResolversParentTypes['User']> = {
  avatar?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  createdAt?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  description?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  email?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  id?: Resolver<ResolversTypes['ID'], ParentType, ContextType>;
  name?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  password?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  siteWeb?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  username?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type Resolvers<ContextType = any> = {
  Mutation?: MutationResolvers<ContextType>;
  Query?: QueryResolvers<ContextType>;
  Token?: TokenResolvers<ContextType>;
  User?: UserResolvers<ContextType>;
};

