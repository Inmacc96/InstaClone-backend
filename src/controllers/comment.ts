import { GraphQLError } from "graphql";
import Comment from "../models/comment";
import User from "../models/user";
import Post from "../models/post";
import { Context } from "../types/Context";
import {
  CommentInput,
  User as UserType,
  Post as PostType,
} from "../types/graphql";

export const addComment = async (input: CommentInput, context: Context) => {
  const { currentUser } = context;
  if (!currentUser)
    throw new GraphQLError("Not authenticated", {
      extensions: {
        code: "UNAUTHENTICATED",
      },
    });

  const { id } = currentUser;

  // Comprobar que el usuario existe
  let user = (await User.findById(id)) as UserType | null;
  if (!user) {
    throw new GraphQLError("User not found", {
      extensions: {
        code: "BAD_USER_INPUT",
      },
    });
  }

  const { idPost, comment } = input;

  try {
    const newComment = new Comment({ idPost, idUser: id, comment });
    await newComment.save();

    const savedComment = await Comment.findById(newComment._id)
      .populate<{ idUser: UserType }>("idUser")
      .populate<{ idPost: PostType }>("idPost");
    if (!savedComment) {
      throw new GraphQLError("Comment not found after saving", {
        extensions: {
          code: "INTERNAL_SERVER_ERROR",
        },
      });
    }
    return savedComment;
  } catch (err) {
    throw new GraphQLError("Error saving the new comment in the database", {
      extensions: {
        code: "INTERNAL_SERVER_ERROR",
      },
    });
  }
};

export const getComments = async (idPost: string) => {
  // Comprobar que existe el post
  let post = await Post.findById(idPost);
  if (!post) {
    throw new GraphQLError("Post not found", {
      extensions: {
        code: "BAD_USER_INPUT",
      },
    });
  }

  // Obtener los comentarios del post
  const comments = await Comment.find({ idPost })
    .populate<{ idUser: UserType }>("idUser")
    .populate<{ idPost: PostType }>("idPost");
  return comments;
};
