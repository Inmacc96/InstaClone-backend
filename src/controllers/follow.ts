import { GraphQLError } from "graphql";
import User from "../models/user";
import Follow from "../models/follow";
import { Context } from "../types/Context";

export const followUser = async (username: string, context: Context) => {
  const { currentUser } = context;

  if (!currentUser)
    throw new GraphQLError("Not authenticated", {
      extensions: {
        code: "UNAUTHENTICATED",
      },
    });

  // Comprobar que el usuario a seguir existe
  const foundUser = await User.findOne({ username });

  if (!foundUser) {
    throw new GraphQLError("User not found", {
      extensions: {
        code: "BAD_USER_INPUT",
        argumentName: "username",
      },
    });
  }

  // Comprobar que el usuario actual no sigue al usuario a seguir
  const alreadyFollowing = await Follow.findOne({
    idUser: currentUser.id,
    follow: foundUser._id,
  });

  if (alreadyFollowing) {
    throw new GraphQLError(`Already follow to ${username}`, {
      extensions: {
        code: "BAD_USER_INPUT",
        argumentName: "username",
      },
    });
  }

  try {
    const follow = new Follow({
      idUser: currentUser.id,
      follow: foundUser._id,
    });
    await follow.save();
  } catch (err) {
    throw new GraphQLError("Error saving the new user in the database", {
      extensions: {
        code: "INTERNAL_SERVER_ERROR",
      },
    });
  }
};

export const isFollowingUser = async (username: string, context: Context) => {
  const { currentUser } = context;

  if (!currentUser)
    throw new GraphQLError("Not authenticated", {
      extensions: {
        code: "UNAUTHENTICATED",
      },
    });

  // Comprobar que el posible usuario seguido existe
  const foundUser = await User.findOne({ username });

  if (!foundUser) {
    throw new GraphQLError("User not found", {
      extensions: {
        code: "BAD_USER_INPUT",
        argumentName: "username",
      },
    });
  }

  // Comprobar si el usuario logeado sigue a username
  const isFollowing = await Follow.findOne({
    idUser: currentUser.id,
    follow: foundUser._id,
  });

  return !!isFollowing;
};
