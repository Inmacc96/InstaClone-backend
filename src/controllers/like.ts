import { GraphQLError } from "graphql";
import { Context } from "../types/Context";
import Like from "../models/like";

export const addLike = async (idPost: string, context: Context) => {
  const { currentUser } = context;

  if (!currentUser)
    throw new GraphQLError("Not authenticated", {
      extensions: {
        code: "UNAUTHENTICATED",
      },
    });

  // Comprobar que el usuario actual no ha dado ya like alpost
  const alreadyLikePost = await Like.findOne({
    idPost,
    idUser: currentUser.id,
  });

  if (alreadyLikePost) {
    throw new GraphQLError(`Already like this post`, {
      extensions: {
        code: "BAD_USER_INPUT",
      },
    });
  }

  try {
    const newLike = new Like({
      idPost,
      idUser: currentUser.id,
    });
    await newLike.save();
  } catch (err) {
    throw new GraphQLError(`Error saving the like`, {
      extensions: {
        code: "INTERNAL_SERVER_ERROR",
      },
    });
  }
};
