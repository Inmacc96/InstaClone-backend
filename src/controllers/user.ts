import bcrypt from "bcryptjs";
import * as dotenv from "dotenv";
import { v2 as cloudinary } from "cloudinary";
import { GraphQLError } from "graphql";
import { createToken } from "../helpers";
import User from "../models/user";
import { Context } from "../types/Context";
import { User as UserType, UserUpdateInput } from "../types/graphql";
import {
  UserInput,
  AuthInput,
  QueryGetUserArgs,
  QueryGenerateUploadUrlArgs,
} from "../types/graphql";
dotenv.config({ path: ".env" });

export const createUser = async (input: UserInput) => {
  const { email, username, password } = input;

  // Revisar si el email está en uso
  const foundEmail = await User.findOne({ email });

  if (foundEmail) {
    throw new GraphQLError("This email is already in use", {
      extensions: {
        code: "BAD_USER_INPUT",
        argumentName: "email",
      },
    });
  }

  // Revisar si el username está en uso
  const foundUsername = await User.findOne({ username });

  if (foundUsername) {
    throw new GraphQLError("This username is already in use", {
      extensions: {
        code: "BAD_USER_INPUT",
        argumentName: "username",
      },
    });
  }

  // Hashear el password
  const salt = await bcrypt.genSalt(10);
  const hashedPassword = await bcrypt.hash(password, salt);

  // Guardar en la base de datos
  try {
    const newUser = new User({ ...input, password: hashedPassword });
    await newUser.save();
    return newUser;
  } catch (err) {
    /*   console.log(err); */
    throw new GraphQLError("Error saving the new user in the database", {
      extensions: {
        code: "INTERNAL_SERVER_ERROR",
      },
    });
  }
};

export const authUser = async (input: AuthInput) => {
  const { email, password } = input;

  // Comprobar que el usuario existe
  const foundUser = await User.findOne({ email });

  // Comprobar que la contraseña es correcta
  const isCorrectPassword =
    !!foundUser && (await bcrypt.compare(password, foundUser.password));

  if (!foundUser || !isCorrectPassword) {
    throw new GraphQLError("Wrong credentials", {
      extensions: {
        code: "BAD_USER_INPUT",
      },
    });
  }
  // Crear el token
  const token = process.env.JWT_SECRET_KEY
    ? createToken(foundUser, process.env.JWT_SECRET_KEY, "24h")
    : null;

  return { token };
};

export const getUser = async ({ id, username }: QueryGetUserArgs) => {
  let user = null;

  if (id) user = await User.findById(id);
  if (username) user = await User.findOne({ username });

  if (!user) {
    throw new GraphQLError("User doesn't exist", {
      extensions: {
        code: "BAD_USER_INPUT",
      },
    });
  }

  return user;
};

export const generateUploadUrl = (
  { folder }: QueryGenerateUploadUrlArgs,
  context: Context
) => {
  const { currentUser } = context;
  if (!currentUser)
    throw new GraphQLError("Not authenticated", {
      extensions: {
        code: "UNAUTHENTICATED",
      },
    });

  const { id } = currentUser;
  const timestamp = Math.round(new Date().getTime() / 1000);

  const uploadParams = {
    folder: `instaclone/${folder}`,
    allowed_formats: ["png", "jpeg"],
    public_id: id,
    timestamp,
  };

  const signature = cloudinary.utils.api_sign_request(
    uploadParams,
    process.env.CLOUDINARY_API_SECRET!
  );

  return { signature, timestamp };
};

export const updateAvatar = async (urlImage: string, context: Context) => {
  const { currentUser } = context;
  if (!currentUser)
    throw new GraphQLError("Not authenticated", {
      extensions: {
        code: "UNAUTHENTICATED",
      },
    });

  const { id } = currentUser;

  // Comprobar que el usuario existe
  let user = (await User.findById(id)) as UserType | null;
  if (!user) {
    throw new GraphQLError("User not found", {
      extensions: {
        code: "BAD_USER_INPUT",
      },
    });
  }

  // Obtener el usuario autenticado y actualizar el valor avatar
  user = (await User.findByIdAndUpdate(
    id,
    { avatar: urlImage },
    { new: true }
  )) as UserType;

  return user;
};

export const deleteAvatar = async (context: Context) => {
  const { currentUser } = context;
  if (!currentUser)
    throw new GraphQLError("Not authenticated", {
      extensions: {
        code: "UNAUTHENTICATED",
      },
    });

  const { id } = currentUser;

  // Comprobar que el usuario existe
  let user = (await User.findById(id)) as UserType | null;
  if (!user) {
    throw new GraphQLError("User not found", {
      extensions: {
        code: "BAD_USER_INPUT",
      },
    });
  }

  // Obtener el usuario autenticado y actualizar el valor avatar
  user = (await User.findByIdAndUpdate(
    id,
    { $unset: { avatar: 1 } },
    { new: true }
  )) as UserType;

  return user;
};

export const updateUser = async (input: UserUpdateInput, context: Context) => {
  const { currentUser } = context;

  if (!currentUser)
    throw new GraphQLError("Not authenticated", {
      extensions: {
        code: "UNAUTHENTICATED",
      },
    });

  const { id } = currentUser;

  // Comprobar que el usuario existe
  let user = (await User.findById(id)) as UserType | null;
  if (!user) {
    throw new GraphQLError("User not found", {
      extensions: {
        code: "BAD_USER_INPUT",
      },
    });
  }

  // Si cambia la contraseña
  if (input.oldPassword && input.newPassword) {
    // Comprobar que la contraseña actual es correcta
    const isCorrectPassword = await bcrypt.compare(
      input.oldPassword,
      user.password
    );
    if (!isCorrectPassword) {
      throw new GraphQLError("Wrong password", {
        extensions: {
          code: "BAD_USER_INPUT",
        },
      });
    }
    // Actualizar la nueva contraseña encriptada
    const salt = await bcrypt.genSaltSync(10);
    const hashedNewPassword = await bcrypt.hash(input.newPassword, salt);

    user = (await User.findByIdAndUpdate(
      id,
      { password: hashedNewPassword },
      { new: true }
    )) as UserType;
  } else {
    // Cambia cualquier otro valor
    user = (await User.findByIdAndUpdate(id, input, { new: true })) as UserType;
  }

  return user;
};
