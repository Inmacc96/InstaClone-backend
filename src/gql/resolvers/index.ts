import { Resolvers } from "../../types/graphql";
import queries from "./queries";
import mutations from "./mutations";

const resolvers: Resolvers = {
  Query: {
    ...queries,
  },
  Mutation: {
    ...mutations,
  },
};

export default resolvers;
