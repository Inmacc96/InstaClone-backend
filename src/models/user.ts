import { Schema, model } from "mongoose";
import { User } from "../types/graphql";

const UserSchema = new Schema<User>({
  name: {
    type: String,
    require: true,
    trim: true,
  },
  username: {
    type: String,
    require: true,
    trim: true,
    unique: true,
  },
  email: {
    type: String,
    require: true,
    trim: true,
    unique: true,
  },
  avatar: {
    type: String,
    trim: true,
  },
  website: {
    type: String,
    trim: true,
  },
  description: {
    type: String,
    trim: true,
  },
  password: {
    type: String,
    require: true,
    trim: true,
  },
  createdAt: {
    type: Date,
    default: Date.now(),
  },
});

export default model<User>("User", UserSchema);
