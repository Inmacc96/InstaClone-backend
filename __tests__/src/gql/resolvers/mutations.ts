import { ApolloServer, BaseContext } from "@apollo/server";
import { readFileSync } from "fs";
import { createUser } from "../../../../src/controllers/user";
import resolvers from "../../../../src/gql/resolvers";
import { UserInput, User } from "../../../../src/types/graphql";

const typeDefs = readFileSync("src/gql/schema.graphql", { encoding: "utf-8" });

jest.mock("../../../../src/controllers/user", () => ({
  createUser: jest.fn(),
}));

const userInput: UserInput = {
  name: "user test",
  username: "usertest",
  email: "usertest@usertest.com",
  password: "usertest123",
};

const userTest: User = {
  id: "ObjectId('6408bddf556f56e7beffa95c')",
  ...userInput,
  password: "$2a$10$eAX2vLXJw/6jpYKx33kOL.RVxZYZ2eeGUcR/3NsQOmK2Q7Ds7QmW2",
};

let testServer: ApolloServer<BaseContext>;

beforeEach(() => {
  jest.resetAllMocks();
  testServer = new ApolloServer<BaseContext>({ typeDefs, resolvers });
});

describe("[Mutation.newUser]", () => {
  const query = `#graphql
    mutation newUser($input: UserInput!){
      newUser(input: $input){
      id
      name
      email
      username
      password
      }
  }
`;

  it("should create a new user", async () => {
    (createUser as jest.Mock).mockResolvedValue(userTest);

    const response = await testServer.executeOperation({
      query,
      variables: { input: userInput },
    });

    expect(createUser).toHaveBeenCalledWith(userInput);
    expect(response).toMatchSnapshot();
  });
});
