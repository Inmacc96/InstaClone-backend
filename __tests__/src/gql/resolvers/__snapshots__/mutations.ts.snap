// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Mutation.newUser] should create a new user 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "newUser": {
          "email": "usertest@usertest.com",
          "id": "ObjectId('6408bddf556f56e7beffa95c')",
          "name": "user test",
          "password": "$2a$10$eAX2vLXJw/6jpYKx33kOL.RVxZYZ2eeGUcR/3NsQOmK2Q7Ds7QmW2",
          "username": "usertest",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
